### PR TITLE
Add combo key press

### LINKED ISSUE #285: Pressing keys in combination is not supported
On current `main` (12eeeb5) the following test:

https://github.com/grafana/xk6-browser/blob/61fb0c26e29d1b95209ab1080e49f08875b8e8d6/tests/keyboard_test.go#L72-L94

... fails with:
```
panic: cannot press key: cannot do key down: "Shift+KeyA" is not a valid key for layout "us" [recovered]
        panic: cannot press key: cannot do key down: "Shift+KeyA" is not a valid key for layout "us"

goroutine 67 [running]:
testing.tRunner.func1.2({0x1850c60, 0xc00059fef0})
        /home/ivan/.local/share/go/src/testing/testing.go:1209 +0x36c
testing.tRunner.func1()
        /home/ivan/.local/share/go/src/testing/testing.go:1212 +0x3b6
panic({0x1850c60, 0xc00059fef0})
        /home/ivan/.local/share/go/src/runtime/panic.go:1047 +0x266
go.k6.io/k6/js/common.Throw(0xc0005d7cc0, {0x1a7c8c0, 0xc0005ba520})
        /home/ivan/Projects/grafana/xk6-browser/vendor/go.k6.io/k6/js/common/util.go:36 +0x79
github.com/grafana/xk6-browser/common.k6Throw({0x1a9a780, 0xc00057c090}, {0x189cd1d, 0x14}, {0xc0005d7de8, 0x1, 0x1})
        /home/ivan/Projects/grafana/xk6-browser/common/helpers.go:214 +0x22f
github.com/grafana/xk6-browser/common.(*Keyboard).Press(0xc0003561e0, {0x188ec4a, 0xa}, {0x0, 0x0})
        /home/ivan/Projects/grafana/xk6-browser/common/keyboard.go:91 +0x1e5
github.com/grafana/xk6-browser/tests.TestKeyboardPress.func4(0x0)
        /home/ivan/Projects/grafana/xk6-browser/tests/keyboard_test.go:81 +0x12e
testing.tRunner(0xc0001f21a0, 0xc0001ee140)
        /home/ivan/.local/share/go/src/testing/testing.go:1259 +0x230
created by testing.(*T).Run
        /home/ivan/.local/share/go/src/testing/testing.go:1306 +0x727
```

This is more of a feature request than a bug, since [Playwright supports this](https://playwright.dev/docs/api/class-keyboard).

### PR DESCRIPTION
**Pull request by ankur22**
This fixes https://github.com/grafana/xk6-browser/issues/285

Shift+X when passed to press() and used in conjunction with down() and up() now will change the casing of the character, as long as the character is in the format `KeyX`. A key press that is `x` or `X` will ignore the shift modifier key. This fix matches the playwright behaviour.

E.g.

```
Shift+KeyA -> A
KeyA -> a
A -> A
a -> a
Shift+A -> A
Shift+a -> a
```

The Shift modifier will only change the case of a character if KeyA - KeyZ is used.

### CODE REVIEW

File: common/keyboard.go
Comment (imiric) [inline common/keyboard.go:127]: I think you should do the `+` split and this key validation in `press()` instead, so that you don't have to repeat it, and `down()`/`up()` can continue handling single keys. It shouldn't be valid to call `down()`/`up()` from JS on a key combo anyway; that is what `press()` is for.

Also, if I'm reading this right, it will accept combos like `A+B+C`, which don't make sense. Maybe we should add validation that there should be only one non-modifier key pressed? Or should we just assume that `A+B+C` equals `ABC`? I guess we can fallback to whatever Playwright accepts, minus the `Shift+B`/`Shift+b` behavior.

**Review by imiric — CHANGES_REQUESTED** [review]
Nicely done, I just think we should move it to `press()` instead.

And re: the `Shift+b` behavior, I don't think Playwright makes sense in that case. `Shift+b` or `Shift+B` should produce `B` in both cases. A lowercase `b` can be typed without `Shift`. It's fine if we diverge with them to improve things. :)

File: common/keyboard.go
Comment (inancgumus) [reply common/keyboard.go:127]: > And re: the `Shift+b` behavior, I don't think Playwright makes sense in that case. `Shift+b` or `Shift+B` should produce `B` in both cases. A lowercase `b` can be typed without `Shift`. It's fine if we diverge with them to improve things. :)

I'm mostly OK with what you propose @imiric, but here are my two cents.

Is diverging PW, in this case, worth it? If I were a tester, I'd at least expect similar behavior.  `Shift+KeyA+B+C` means holding down the shift key while pressing the other keys. For example: `Shift+KeyA+KeyB+B+b` is `ABBb`, which makes sense to me.

Puppeteers also work with `KeyA`, `KeyX`, etc. style.

So instead of `Shift+B`, it should be `Shift+KeyB`, not `Shift+B`. `KeyX` is also in tandem with JS key codes: https://javascript.info/keyboard-events. As I said, I'm also OK with what you proposed.

**Comment by inancgumus** [comment]
> And re: the `Shift+b` behavior, I don't think Playwright makes sense in that case. `Shift+b` or `Shift+B` should produce `B` in both cases. A lowercase `b` can be typed without `Shift`. It's fine if we diverge with them to improve things. :)

I'm mostly OK with what you propose @imiric, but here are my two cents.

Is diverging PW, in this case, worth it? If I were a tester, I'd at least expect similar behavior.  `Shift+KeyA+B+C` means holding down the shift key while pressing the other keys. For example: `Shift+KeyA+KeyB+B+b` is `ABBb`, which makes sense to me.

Puppeteers also work with `KeyA`, `KeyX`, etc. style.

So instead of `Shift+B`, it should be `Shift+KeyB`, not `Shift+B`. `KeyX` is also in tandem with JS key codes: https://javascript.info/keyboard-events.

As I said, I'm also OK with what you proposed. But, I'm not sure whether we're improving things or making people's life harder 😅

**Review by inancgumus — COMMENTED** [review]
As @imiric also noticed, splitting a given key should be done in `press`. `down` and `up` are only responsible for a single key, and that key can be a modifier as well. For more info: https://github.com/grafana/xk6-browser/pull/326/files#r885768387.

**Comment by ankur22** [comment]
I'm unable to get the `Meta` and `Alt` modifier keys to work with any of the existing functions e.g.

```go
		kb.Press("Shift+KeyA", nil)
		kb.Press("Shift+b", nil)
		kb.Press("Shift+C", nil)

		metaKey := "Control"
		if runtime.GOOS == "darwin" {
			metaKey = "Meta"
		}
		kb.Press(metaKey+"+A", nil)
		kb.Press("Delete", nil)
```

OR

```go
		kb.Press("Shift+KeyA", nil)
		kb.Press("Shift+b", nil)
		kb.Press("Shift+C", nil)

		kb.Press("Alt+ArrowLeft", nil)
		kb.Press("Delete", nil)
```

This should delete the text that has been written into the input field. To continue to track and resolve this a new ticket has been opened: #?

File: common/keyboard.go
Comment (ankur22) [reply common/keyboard.go]: Done

File: common/keyboard.go
Comment (inancgumus) [inline common/keyboard.go:93]: Nit: No need to put an empty line here, but your call :)

File: common/keyboard.go
Comment (inancgumus) [inline common/keyboard.go:261]: This timer will leak if `t.C` matches.

There are similar leaks in the code (I guess we missed them before). Maybe it's time to fix them all.

File: common/keyboard.go
Comment (inancgumus) [inline common/keyboard.go:265]: I couldn't understand what is this for? Should it be `strings.Contains`?

File: common/keyboard.go
Comment (inancgumus) [inline common/keyboard.go:274]: Is there a reason why this is not a `range` as the above `down` loop?

**Review by inancgumus — CHANGES_REQUESTED** [review]
I'll continue looking into this tomorrow, but it failed on my machine as follows:

```
--- FAIL: TestPageInputSpecialCharacters (0.68s)
    page_test.go:341: 
                Error Trace:    page_test.go:341
                Error:          Not equal: 
                                expected: "test@k6.io"
                                actual  : "test@k^>io"
                            
                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -test@k6.io
                                +test@k^>io
                Test:           TestPageInputSpecialCharacters
    page_test.go:341: 
                Error Trace:    page_test.go:341
                Error:          Not equal: 
                                expected: "<hello WoRlD \\/>"
                                actual  : "<hello WoRlD |?>"
                            
                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -<hello WoRlD \/>
                                +<hello WoRlD |?>
                Test:           TestPageInputSpecialCharacters
    page_test.go:341: 
                Error Trace:    page_test.go:341
                Error:          Not equal: 
                                expected: "¯\\_(ツ)_/¯"
                                actual  : "¯|_(ツ)_?¯"
                            
                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -¯\_(ツ)_/¯
                                +¯|_(ツ)_?¯
                Test:           TestPageInputSpecialCharacters
```

Update: CI has reported the failure as well: https://github.com/grafana/xk6-browser/runs/7114936484?check_suite_focus=true

File: common/keyboard.go
Comment (ankur22) [reply common/keyboard.go:274]: It's going backwards:

a->b->c
c->b->a

File: common/keyboard.go
Comment (ankur22) [reply common/keyboard.go:265]: Yeah, this is an interesting edge case, so in playwright:

1. `press("+")` -> types `+`
2. `press("++")` -> errors with `Unknown key: ""`
3. `press("+++")` -> types `++`
4. `press("Sift++")` -> types `+`

I'm trying to code the same behaviour, any ideas how this could be done?

File: common/keyboard.go
Comment (inancgumus) [reply common/keyboard.go:285]: The timer creates a goroutine, but `t.Stop` is only called to stop the goroutine in the first `case`.

File: common/keyboard.go
Comment (inancgumus) [inline common/keyboard.go:317]: The linter is right here. It's better to convert the if-else chain into a switch.

```go
for _, k := range keys {
	switch {
	case k == '+' && ss != "":
		...
...
```

And, you can use [`strings.Builder`](https://stackoverflow.com/questions/1760757/how-to-efficiently-concatenate-strings-in-go/47798475#47798475) to append runes without converting them.

File: common/keyboard.go
Comment (inancgumus) [inline common/keyboard.go:46]: You can put `keyAToKeyZ` into a function and ask the function instead of this `nolint` directive. The Go compiler will most probably inline the function.

Besides, I'm not sure whether we need the map. You can simply create a function and return `true`/`false` depending on the pattern `Key[A-Z]`.

**Review by inancgumus — CHANGES_REQUESTED** [review]
I looked more into this. Here are my new findings.

File: common/keyboard.go
Comment (inancgumus) [inline common/keyboard.go:253]: Can you move this logic (`(ok || foundInShift) && shift != 0 && srcKeyDef.ShiftKey != "")`) into a closure? It's harder to see what it does.

File: common/keyboard.go
Comment (ankur22) [reply common/keyboard.go:285]: Ok, as far as i can tell, the timer doesn't need to be stopped since we've read off of the channel. Although we should be reading off of the channel when we call `Stop` in the case above. Sound correct to you? Source: https://pkg.go.dev/time#NewTimer

File: common/keyboard.go
Comment (ankur22) [reply common/keyboard.go]: Yeah it is the same for PW, here's the script I used to test this:

```js
// @ts-check
const { test, chromium } = require('@playwright/test');

test.describe('Editing', () => {
  test('Type login creds', async () => {
    const browser = await chromium.launch({ headless: false, slowMo: 500 });
    const page = await browser.newPage();
    await page.goto('https://www.bing.com/');
    var elm = await page.$('input[name="q"]');

    await elm.press("+++"); // types in ++
    await elm.press("+"); // types in +
    await elm.press("++"); // will error and exit with `Unknown key: ""`

    await new Promise(resolve => setTimeout(resolve, 5000));

    await browser.close();
  });
});
```

Comment (inancgumus) [reply common/keyboard.go]: Interesting, thanks!

File: common/keyboard.go
Comment (ankur22) [reply common/keyboard.go:295]: Yeah sure, nice little trick 🙂

File: common/keyboard.go
Comment (ankur22) [reply common/keyboard.go:317]: Nice, I hadn't realised that it was idiomatic to write them as switch statements. Source: https://go.dev/doc/effective_go#switch

File: common/keyboard.go
Comment (inancgumus) [reply common/keyboard.go:285]: You're right, sorry 🤦

Comment (inancgumus) [reply common/keyboard.go:285]: Ah, another idea. Why don't we just create a child timeout context? And block there?

```go
if opts.Delay != 0 {
	ctx, _ := context.WithTimeout(k.ctx, time.Duration(opts.Delay) * time.Millisecond)
	<-ctx.Done()
}
```

File: common/keyboard.go
Comment (inancgumus) [reply common/keyboard.go]: Isn't it more like this?

```go
func split(keys string) []string {
	var (
		kk = make([]string, 0, len(keys))
		s  strings.Builder
	)
	for _, r := range keys {
		if r == '+' && s.Len() > 0 {
			kk = append(kk, s.String())
			s.Reset()
		} else {
			s.WriteRune(r)
		}
	}
	kk = append(kk, s.String())

	return kk
}
```

See: https://go.dev/play/p/fJoqZIMTBF5

File: common/keyboard.go
Comment (inancgumus) [inline common/keyboard.go:362]: Isn't it more like this?

```go
func split(keys string) []string {
	var (
		kk = make([]string, 0, len(keys))
		s  strings.Builder
	)
	for _, r := range keys {
		if r == '+' && s.Len() > 0 {
			kk = append(kk, s.String())
			s.Reset()
		} else {
			s.WriteRune(r)
		}
	}
	kk = append(kk, s.String())

	return kk
}
```

See: https://go.dev/play/p/fJoqZIMTBF5

Comment (ankur22) [reply common/keyboard.go:362]: Nice! This is much easier to understand. Can you link me to the PW code please?

File: common/keyboard.go
Comment (imiric) [reply common/keyboard.go]: That trick is not intuitive at all IMO. The previous C-style `for` was clear and straightforward.

Please let's not point out style issues in reviews, unless the reviewer can point out an objective benefit of the change.

**Review by imiric — CHANGES_REQUESTED** [review]
Nice work :+1: Only the test comment seems like a blocker to me, the others are minor suggestions.

It's strange that meta keys don't seem to work. That's a pretty important issue, but we can tackle it later.

Also, remember to mark the PR "Ready for review" when it's done being a draft.

File: common/keyboard.go
Comment (ankur22) [reply common/keyboard.go]: Resolved in: 8b684872b7bdf97135eae6fffd8074c13f16d6b8

File: common/keyboard.go
Comment (ankur22) [reply common/keyboard.go]: Resolved in: 20908e3c5310b6392d96a9768fd84bc1c12b97d0

File: common/keyboard.go
Comment (ankur22) [reply common/keyboard.go]: Resolved in: 8ad3b446b98b3f1c69ed7a43000d8aebc56b3828

File: tests/keyboard_test.go
Comment (ankur22) [reply tests/keyboard_test.go:104]: I've tried to refactor a couple of the tests to unit tests. Some of them I can't really move to unit tests even though they're kind of related to split or not being split.

Resolved in: 71ba9f56f23908eceda6bd93e13911489f93f51f

**Review by imiric — APPROVED** [review]
LGTM, nice work :+1:

I would just move `split()` to `common/helpers.go` and its tests to `helpers_test.go`. On second thought, it might be a uniquely keyboard helper function, so it makes sense where it is. Up to you.

The CI failures don't seem to be related to these changes, though that [`TestPageGotoWaitUntilDOMContentLoaded`](https://github.com/grafana/xk6-browser/runs/7636154407?check_suite_focus=true) is a new one I haven't seen before. :confused: Please squash this into a single commit and push it so the tests run again before merging.

**Pull request merged**
